lib/core/include/dstream.hpp

```diff
@@ -163,6 +230,26 @@ namespace irods::experimental::io
             return transport_->file_descriptor();;
         }
 
+        root_resource_name root_resource_name() const
+        {
+            return transport_->resource_name();
+        }
+
+        std::string resource_hierarchy() const
+        {
+            return transport_->resource_hierarchy();
+        }
+
+        replica_number replica_number() const
+        {
+            return transport_->replica_number();
+        }
+
+        replica_token replica_token() const
+        {
+            return transport_->replica_token();
+        }
+
```

Comment:
these should probably be returned by reference.  how expensive is a copy here, though?

Reply:
allocating a std::string could potentially be expensive.

Reply:
True. I think there are a lot of things to consider here.

The values returned are constructed on demand making it difficult to impossible to return a reference. There is also the case of how often will these functions be invoked. The results will likely be short strings more often than not so they should hit the SBO and avoid heap allocation. These may also hit RVO too.

With that, i feel that the copies here aren't that bad (they may be optimized away).

This is under the experimental namespace too, so we have that going for us :)